Rework clean_merge_asof func

katie-lamb opened the pull request:
I added the helper function `mixed_temporal_gran_merge` that is a rework of `clean_merge_asof`.

Originally used Pandas `resample` - really slow, adopted CG approach from WIP PR.  
This new `mixed_temporal_gran_merge` function takes about half the time of `clean_merge_asof` and can do ALL THE MERGES (not just left) as it's just a wrapper on `merge`. It's still pretty slow though.

*Structure*
- Explode the datetime column into separate `year`, `quarter`, `month`, `day` columns
- Get the list of columns to merge on based on the granularity of the less granular dataframe
- Merge on these columns as well as any additional columns to merge on

*To Do*

- [x] Recreate `report_date` column
- [x] Break timeseries enumeration into a separate function, then make a higher level function that coordinates the two
- [x] Write out test cases
- [x] Change `lesser_freq` argument into `on_cols` with the temporal columns included
- [x] Make test cases into unit tests
- [x] Change out the `clean_merge_asof` calls for `mixed_temporal_gran_merge`

*Current Shortcomings*
- I've tried some edge cases but so far only with the `generation_fuel_eia923` data and the `plant_utils_eia860` data. I was a little confused about how the specifics of the mcoe `clean_merge_asof` calls so I haven't changed them to `mixed_temporal_gran_merge` yet.
- The function doesn't enumerate all of the possible dates in the date range. For example the annual dataframes aren't broadcast to all 12 months, they're just split into data points for January for each year.

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,107 @@ def is_doi(doi):
+    # this fixes an off by one issue with upsampling
+    if report_time_in_period == "first":
+        extra_year_df = (
+            right.groupby(on_cols)[right_date_col].max() + pd.offsets.YearBegin()
```
I need to change this so the offset (`pd.offsets.YearBegin()` for now) change based on the granularity of the right dataframe. Maybe I'll check what the offset is between two rows and use this time delta as the offset.

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,108 @@ def is_doi(doi):
+    right = pd.concat([right, extra_year_df])
+    # upsample to higher granularity
+    if report_time_in_period == "first":
+        right = right.set_index(right_date_col).groupby(on_cols).resample(gran).ffill()
```
I'm not entirely sure why this `resample` call is so slow. I guess upsampling just takes a while. There are a lot of limitations of `resample` which makes it necessary to call `set_index` and then `groupby`. 

[See this post](https://stackoverflow.com/questions/68604681/upscalling-multiindex-dataframe-in-pandas-1-3)

I think this `resample` needs to go

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,141 @@ def is_doi(doi):
+    merge_on = [col for col in right if col in all_date_cols] + on_cols
+    out = left.merge(right, on=merge_on, how=merge_type)
+
+    """
```
Left all this in for now commented out. It's old.

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,227 @@ def is_doi(doi):
     return bool(re.match(doi_regex, doi))
 
 
+def full_timeseries_temporal_merge(
```
Is there a better way for this higher level function to accept all these arguments?

zaneselvans replied:
It looks like `full_timeseries_temporal_merge` isn't being called anywhere in the code right now. Is that expected? Was @cmgosnell using it somewhere else in the RMI / PPL code?

cmgosnell replied:
I want to use it in line with what I was trying to do in [this PR](https://github.com/catalyst-cooperative/pudl/pull/1478). I assume we can scrap that PR and insert this new merge into `pudl.analysis.allocate_net_gen.associate_generator_tables`.

I can also imagine it being the right thing to use in `pudl.analysis.mcoe.mcoe()`. The first merge in there uses an arg `all_gens` to determine the left or right-ness of the merge but if we're using an monthly `pudl_out` the output will include a ton of 1/1/year generator records for those that don't show up in the monthly data.

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,227 @@ def is_doi(doi):
+    right_date_col: str = "report_date",
+    new_date_col: str = "report_date",
+    shared_merge_cols: List[str] = [],
+    temporal_merge_cols: List[str] = ["year"],
```
Can you constrain the values that can be in a list, e.g. "year", "month", "quarter", "day"? It's a list not a string so `Literal` won't work?

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,227 @@ def is_doi(doi):
+    else:
+        suffixes = ["_x", "_y"]
+    cols_to_drop = ["year", "month", "day"]
+    if "quarter" in out:
```
"Quarter" columns are treated differently because they're not used to reconstruct a date.

katie-lamb replied:
Yep, that would be much simpler. Will try that. I've been looking at this too long and it got too complex.

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,227 @@ def is_doi(doi):
+    merge_cols = temporal_merge_cols + shared_merge_cols
+    out = left.merge(right, on=merge_cols, how=merge_type, **kwargs)
+
+    # reconstruct the report_date column and do some column cleanup
```
This reconstruction of `report_date` part isn't the cleanest. Basically if there is an x and y column for each of year, month, day, I took the larger value of the x and y sides as the more granular date (because the less granular date has the same value or 1 (january, first of month) if reported at the beginning of the period).

katie-lamb commented:
I think these tests cover all the cases in the check list of cases from the above comment. I still haven't switched out the `clean_merge_asof` calls for `mixed_temporal_gran_merge` calls. I figured I'd wait to see if there needs to be any major changes.

katie-lamb commented on test/unit/analysis/mixed_temporal_merge_test.py:
```diff
@@ -0,0 +1,358 @@
+"""Tests for merging datasets at different temporal frequencies."""
```
Should these tests go in `unit/helpers` instead of `unit/analysis`?

katie-lamb commented on src/pudl/analysis/mcoe.py:
```diff
@@ -220,13 +202,12 @@ def fuel_cost(pudl_out):
-        by=["plant_id_eia", "generator_id"],
+        shared_merge_cols=["plant_id_eia", "generator_id"],
+        temporal_merge_cols=["year"],
+        merge_type="left",
```
The comment that was here for the call to `clean_merge_asof` says we're inner merging, but aren't we actually left joining? I made the merge a left join but I think inner and left join essentially do the same thing here?

zaneselvans commented on test/unit/analysis/mixed_temporal_merge_test.py:
```diff
@@ -0,0 +1,358 @@
+from pandas.tseries.offsets import BYearEnd
+
+import pudl.helpers
+
```
The `quarter` case is a little unusual, not being used currently, and also not being tested. It seems like we should either test it, or remove it (and add it in when we need to for the FERC data).

katie-lamb replied:
I put in a place holder test for merging quarterly data onto monthly data but this should be updated to use FERC data for testing when we need to do this merge for the FERC data.

zaneselvans reviewed (changes requested):
I'm having a hard time reading the unit test cases, I think partly because they don't seem to correspond to real merge cases in the data that I'm familiar with. And reading inline DataFrame definitions is always hard.

zaneselvans commented:
I ran the data validations locally, and all of the failures I got were due to the number of rows being off by a little bit. The biggest discrepancy I had was only 0.144%:

```
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-bf_eia923-1309942-1309942-109807] - ValueError: bf_eia923: found 1309908 rows, expected 1309942. Off by -0.003%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-frc_eia923-560377-230063-22686] - ValueError: frc_eia923: found 560360 rows, expected 560377. Off by -0.003%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-gen_eia923-559570-559570-46718] - ValueError: gen_eia923: found 559546 rows, expected 559570. Off by -0.004%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-gf_eia923-2507870-2507870-214709] - ValueError: gf_eia923: found 2507830 rows, expected 2507870. Off by -0.002%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-gf_nonuclear_eia923-2492481-2492481-213422] - ValueError: gf_nonuclear_eia923: found 2492441 rows, expected 2492481. Off by -0.002%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-bf_eia923-1309942-1309942-109807] - ValueError: bf_eia923: found 109773 rows, expected 109807. Off by -0.031%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-frc_eia923-560377-230063-22686] - ValueError: frc_eia923: found 22679 rows, expected 22686. Off by -0.031%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-gen_eia923-559570-559570-46718] - ValueError: gen_eia923: found 46694 rows, expected 46718. Off by -0.051%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-gf_eia923-2507870-2507870-214709] - ValueError: gf_eia923: found 214669 rows, expected 214709. Off by -0.019%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-gf_nonuclear_eia923-2492481-2492481-213422] - ValueError: gf_nonuclear_eia923: found 213382 rows, expected 213422. Off by -0.019%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-hr_by_unit-341681-28560] - ValueError: hr_by_unit: found 28519 rows, expected 28560. Off by -0.144%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-hr_by_gen-514647-42963] - ValueError: hr_by_gen: found 42935 rows, expected 42963. Off by -0.065%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-fuel_cost-514642-42958] - ValueError: fuel_cost: found 42935 rows, expected 42958. Off by -0.054%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-capacity_factor-559570-46718] - ValueError: capacity_factor: found 46694 rows, expected 46718. Off by -0.051%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-mcoe-559618-46718] - ValueError: mcoe: found 46694 rows, expected 46718. Off by -0.051%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-bf_eia923-1309942-1309942-109807] - ValueError: bf_eia923: found 1309908 rows, expected 1309942. Off by -0.003%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-frc_eia923-560377-230063-22686] - ValueError: frc_eia923: found 230057 rows, expected 230063. Off by -0.003%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-gen_eia923-559570-559570-46718] - ValueError: gen_eia923: found 559546 rows, expected 559570. Off by -0.004%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-gf_eia923-2507870-2507870-214709] - ValueError: gf_eia923: found 2507830 rows, expected 2507870. Off by -0.002%, allowed margin of 0.000%
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-gf_nonuclear_eia923-2492481-2492481-213422] - ValueError: gf_nonuclear_eia923: found 2492441 rows, expected 2492481. Off by -0.002%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-hr_by_unit-341681-28560] - ValueError: hr_by_unit: found 341640 rows, expected 341681. Off by -0.012%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-hr_by_gen-514647-42963] - ValueError: hr_by_gen: found 514619 rows, expected 514647. Off by -0.005%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-fuel_cost-514642-42958] - ValueError: fuel_cost: found 514619 rows, expected 514642. Off by -0.004%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-capacity_factor-559570-46718] - ValueError: capacity_factor: found 559546 rows, expected 559570. Off by -0.004%, allowed margin of 0.000%
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-mcoe-559618-46718] - ValueError: mcoe: found 559594 rows, expected 559618. Off by -0.004%, allowed margin of 0.000%
```

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,229 @@ def is_doi(doi):
+    """
+    Merge dataframes with different date frequencies and expand to a full timeseries.
+
+    Arguments: see arguments for ``mexed_temporal_gran_merge`` and ``expand_timeseries``
```
haha, good catch

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,229 @@ def is_doi(doi):
+    """
+
+    def separate_date_cols(df, date_col_name):
+        df[date_col_name] = pd.to_datetime(df[date_col_name])
```
I decided to convert the column to a string if it's not not already a `datetime` column as well as warn if the type is integer. Checking if it's a string or an integer doesn't work that well because if it's either all strings or it's mixed type (some strings and some integers) then it will be read as an `object`, so it's safest to just convert to a string.

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,229 @@ def is_doi(doi):
+        df.loc[:, "year"] = df[date_col_name].dt.year
+        df.loc[:, "quarter"] = df[date_col_name].dt.quarter
+        df.loc[:, "month"] = df[date_col_name].dt.month
+        df.loc[:, "day"] = df[date_col_name].dt.day
```
I changed the column names to have the suffix `_temp_for_merge` and also added a warning if one of the columns is already in a dataframe.

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,229 @@ def is_doi(doi):
+        df = df.drop(date_col_name, axis=1)
+        return df
+
+    right = right.copy()
+    left = left.copy()
```
This was leftover from when I was testing. Probably best to take out.

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,229 @@ def is_doi(doi):
+        ),
+        columns=[date_col],
+    )
+    ind_cols = [date_range[date_col]] + [df[id_col].unique() for id_col in id_cols]
```
Good catch, you're right, it should only include every observed combination of non date primary key columns, not every possible combination. I believe I fixed this and added some more detailed comments

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -339,6 +339,229 @@ def is_doi(doi):
+    date_col: str = "report_date",
+    start: str = None,
+    end: str = None,
+    freq="MS",
```
> Should `freq` be restricted to a particular set of literals? Can we get that set from `timeseries.offset_aliases`?

Since you can combine frequencies and still have a valid string ("5day5min") it's hard to restrict `freq` to a set of literals. I instead try/except with the pandas function `to_offset(freq)` which will generate an exception if `freq` is not valid.

> It seems like this function implicitly assumes that all dates will be "start" format -- beginning of the month/year etc. Might want to state that explicitly. E.g. if you only do `ffill` to fill in the data values it'll work as expected one `MS` data, but it'll do nothing with month-end data (which would require `bfill`)

I believe it still works with "end" format dates in certain cases. Take the following input for example with `frequency = "M"` and a date range from `2019-01-01` to `2019-05-01`: 
```
plant_id    report_date    data
1           2019-01-31       1
1           2019-04-30       2 
```
This would `ffill` to 
```
plant_id    report_date    data
1           2019-01-31       1
1           2019-02-28       1
1           2019-03-31       1
1           2019-04-30       2 
1           2019-05-31       2 
```
and would `bfill` to
```
plant_id    report_date    data
1           2019-01-31       1
1           2019-02-28       2
1           2019-03-31       2
1           2019-04-30       2 
1           2019-05-31       Nan 
```
It depends on the use case but I think `ffill` is more intuitive to me.

There's also the case where you'd want to use `bfill`, but I think this is handled already in `date_merge`. For example, if you have annual data reported on December, 31st being merged with monthly data. Specifying `report_at_start = False` in `date_merge` would result in the newly merged on annual records being backfilled for every month in that year. If some months were not observed in the monthly dataset, then you could then call `expand_timeseries` to fill in these missing months.

katie-lamb replied:
> Also maybe want to alert the user to the need to explicitly indicate the end date at the end of the year if they're working with annual data... like you want `start="2019-01-01"` and `end="2020-12-31"` -- the default args would get the start date right, but only take you up to January in the last year. Right?
> 
In this case since the data is annual wouldn't you only want data for `2019-01-01` and `2020-01-01`, not for `2021-01-01`? And if the data was monthly instead of annual you'd get data through December, 2020. 

> And some warning against using this function for irregularly reported data. It can turn annual data into corresponding homogeneous monthly data, but if you apply it to something like the `fuel_receipts_costs_eia923` table you'll get meaningless output.

Yeah, irregularly reported data is an issue unless you had a very granular frequency, like daily or something. But this seems like more of user error. For example, if you had data reported on the following dates:
```
2019-01-06
2019-02-03
2019-02-06
```
then it seems like you wouldn't want to actually give it a frequency of `M` or `MS`, since you probably want it at a daily resolution? If your dates were at a regular interval but on the 6th day of the month or something, you can still pass a frequency string corresponding to monthly starting on the 6th day.

katie-lamb commented:
I think I addressed a bunch of @zaneselvans review comments. There are still a few questions left: 
- Is there a way to make some of these unit test cases more clear? Or just formatting them in some way so you don't just need to read the inline DataFrame definitions? I think I need more context (maybe from @cmgosnell ) on how some of these merges are being used so that the test data can be more real. Also I'm not sure if there's already an instance (maybe @cmgosnell wanted this for the plants parts list) where we want to merge and then fill in the full time series, but currently it's not being called.
- I added a test for merging quarterly data onto monthly data, but this should probably be updated once the FERC data is actually using this type of quarterly merge. It's kind of acting as a placeholder right now - should I just take it out?
- I think there's still some confusion on how `expand_timeseries` should work (see my replies to [this comment](https://github.com/catalyst-cooperative/pudl/pull/1550#discussion_r846389376)). If this doesn't make sense, let's talk more about it.

cmgosnell commented:
hi y'all! Sorry for being late to the party @katie-lamb . I thought i had left a comment before on @zaneselvans 's question about why the `full_timeseries_temporal_merge` isn't being used.... but alas I'll do it again.

I believe there are two places where the `full_timeseries_temporal_merge` should be implemented:
* in `pudl.analysis.mcoe.mcoe()` when the `all_gens` arg is set to True.
```
mcoe_out = (
        pd.merge(
            left=(
                pudl_out.gens_eia860()
                .assign(year=lambda x: x.report_date.dt.year)
                .drop("report_date", axis="columns")
            ),
            right=mcoe_out.assign(year=lambda x: x.report_date.dt.year),
            # This "how" determines whether MCOE or gens_eia860 is the backbone
            how="left" if all_gens else "right",
            on=["year", "plant_id_eia", "generator_id"],
        )
```
The bulk of the data in the mcoe output comes from monthly reported 923 tables, which only cover a subset of the total generators reported to other 923 tables and 860 tables. Because the mcoe output is the most complete output in terms of containing the most calculated data points about generators, it is often used as the starting place for analysis about generators. But it can be helpful to have all of the generators in there. Right now when `pudl_out.freq = "MS"` the output includes a bunch of monthly records from `mcoe_out` and a bunch of annual records from `pudl_out.gens_eia860()` - which is weird. I think we could use `full_timeseries_temporal_merge()` when `all_gens = True` to return an output with a consistent time granularity (i.e. ALL of the generators are in monthly records regardless of whether that generator ever had any monthly reported information.)
* (this one can be implemented later!) I originally was trying something like this in a less clean way in this PR https://github.com/catalyst-cooperative/pudl/pull/1478. I would be happy to take whatever you do here and implement that when this PR gets merged in.

katie-lamb commented on src/pudl/analysis/mcoe.py:
```diff
@@ -475,23 +458,16 @@ def mcoe(
+        start=pudl_out.start_date,
+        end=pudl_out.end_date,
+        freq=pudl_out.freq,
+    ).pipe(pudl.validate.no_null_rows, df_name="mcoe_all_gens", thresh=0.9)
```
@cmgosnell I changed this based on your [comment here](https://github.com/catalyst-cooperative/pudl/pull/1550#issuecomment-1106597818). Let me know if the parameters feel right here. I changed this so if `all_gens` is True an `outer` merge is performed instead of a `left` merge (using the keys from `gens_eia860`. I think they do the same thing since I was interpreting it as the generators in `mcoe_out` are a subset of the generators in `gens_eia860()`, but it intuitively made more sense to me to do an `outer` merge so that all the generators are included.

zaneselvans commented:
Hey @katie-lamb all of the tests are failing in the middle of the output tests with a `kill -9`, which IIRC is what happens when it runs out of memory (of which a max of 7GB is available on the test runner).

Can the output tests really be taking more than 7GB of memory even with just a single year of data? Is there a change here that we would expect to blow up the memory usage / number of rows dramatically? Or might we have been getting close to the limit gradually and just now tipped over the edge?

zaneselvans commented:
Running `tox -e integration` locally, it looks like I'm getting 10-15GB of memory freed up as soon as it finishes with the MCOE tests, which feels like a lot more than I would expect for a single year of data.  And if it really is the amount for a single year of data, then it would become impossible to run the MCOE for all of the years of data.

zaneselvans commented:
Attempting to run `pudl_out.mcoe()` locally on all years of data, it also uses like 15-20GB of memory, and then eventually it fails an assertion due to the prevalence of (largely) empty rows in the MCOE output, a check we originally put in place to catch bad merges.

zaneselvans commented:
Disabling the check for (mostly) null rows and re-running (using the cached versions of the dataframes which go into building the MCOE) it maxes out my laptop's memory (I have 24GB but some other stuff running too) and then also uses 15GB of swap. The resulting dataframe has about 9 million rows and takes 22.3GB of memory.
```py
mcoe_ms = pudl_out_ms.mcoe()
mcoe_ms.info(show_counts=True, memory_usage="deep", max_cols=150)
```
```
<class 'pandas.core.frame.DataFrame'>
Int64Index: 8982792 entries, 0 to 8982791
Data columns (total 119 columns):
 #    Column                                   Non-Null Count    Dtype         
---   ------                                   --------------    -----         
 0    plant_id_eia                             8982792 non-null  Int64         
 1    generator_id                             8982792 non-null  string        
 2    report_date                              8982792 non-null  datetime64[ns]
 3    unit_id_pudl                             664003 non-null   Int64         
 4    plant_id_pudl                            6536928 non-null  Int64         
 5    plant_name_eia                           6536712 non-null  string        
 6    utility_id_eia                           6535800 non-null  Int64         
 7    utility_id_pudl                          6536928 non-null  Int64         
 8    utility_name_eia                         6536928 non-null  string        
 9    associated_combined_heat_power           6418164 non-null  boolean       
 10   balancing_authority_code_eia             5608032 non-null  string        
 11   balancing_authority_name_eia             5517504 non-null  string        
 12   bga_source                               664003 non-null   string        
 13   bypass_heat_recovery                     5663316 non-null  boolean       
 14   capacity_factor                          695329 non-null   float64       
 15   capacity_mw                              6535608 non-null  float64       
 16   carbon_capture                           1456224 non-null  boolean       
 17   city                                     6203436 non-null  string        
 18   cofire_fuels                             1257408 non-null  boolean       
 19   county                                   6026232 non-null  string        
 20   current_planned_operating_date           1566432 non-null  datetime64[ns]
 21   data_source                              4760313 non-null  string        
 22   deliver_power_transgrid                  4518795 non-null  boolean       
 23   distributed_generation                   5238588 non-null  boolean       
 24   duct_burners                             6232536 non-null  boolean       
 25   energy_source_1_transport_1              3913128 non-null  string        
 26   energy_source_1_transport_2              123144 non-null   string        
 27   energy_source_1_transport_3              5016 non-null     string        
 28   energy_source_2_transport_1              1529280 non-null  string        
 29   energy_source_2_transport_2              21576 non-null    string        
 30   energy_source_2_transport_3              2556 non-null     string        
 31   energy_source_code_1                     6535188 non-null  string        
 32   energy_source_code_2                     1817112 non-null  string        
 33   energy_source_code_3                     373138 non-null   string        
 34   energy_source_code_4                     129912 non-null   string        
 35   energy_source_code_5                     62976 non-null    string        
 36   energy_source_code_6                     26148 non-null    string        
 37   ferc_cogen_status                        6415596 non-null  boolean       
 38   ferc_exempt_wholesale_generator          6365700 non-null  boolean       
 39   ferc_small_power_producer                6372504 non-null  boolean       
 40   fluidized_bed_tech                       62988 non-null    boolean       
 41   fuel_cost_from_eiaapi                    497588 non-null   object        
 42   fuel_cost_per_mmbtu                      276231 non-null   float64       
 43   fuel_cost_per_mwh                        271257 non-null   float64       
 44   fuel_type_code_pudl                      6535188 non-null  string        
 45   fuel_type_count                          6536928 non-null  float64       
 46   grid_voltage_2_kv                        285180 non-null   float64       
 47   grid_voltage_3_kv                        43488 non-null    float64       
 48   grid_voltage_kv                          5734356 non-null  float64       
 49   heat_rate_mmbtu_mwh                      611550 non-null   float64       
 50   iso_rto_code                             3378636 non-null  string        
 51   latitude                                 6064104 non-null  float64       
 52   longitude                                6142668 non-null  float64       
 53   minimum_load_mw                          1985496 non-null  float64       
 54   multiple_fuels                           4472930 non-null  boolean       
 55   nameplate_power_factor                   2149656 non-null  float64       
 56   net_generation_mwh                       707732 non-null   float64       
 57   operating_date                           5856696 non-null  datetime64[ns]
 58   operating_switch                         14040 non-null    string        
 59   operational_status                       6535704 non-null  string        
 60   operational_status_code                  6535644 non-null  string        
 61   original_planned_operating_date          1563540 non-null  datetime64[ns]
 62   other_combustion_tech                    83484 non-null    boolean       
 63   other_modifications_date                 98208 non-null    datetime64[ns]
 64   other_planned_modifications              1233292 non-null  boolean       
 65   owned_by_non_utility                     3876504 non-null  boolean       
 66   ownership_code                           6526704 non-null  string        
 67   planned_derate_date                      27360 non-null    datetime64[ns]
 68   planned_energy_source_code_1             27948 non-null    string        
 69   planned_modifications                    3497813 non-null  boolean       
 70   planned_net_summer_capacity_derate_mw    16164 non-null    float64       
 71   planned_net_summer_capacity_uprate_mw    93612 non-null    float64       
 72   planned_net_winter_capacity_derate_mw    25152 non-null    float64       
 73   planned_net_winter_capacity_uprate_mw    82116 non-null    float64       
 74   planned_new_capacity_mw                  17244 non-null    float64       
 75   planned_new_prime_mover_code             22932 non-null    string        
 76   planned_repower_date                     32460 non-null    datetime64[ns]
 77   planned_retirement_date                  263992 non-null   datetime64[ns]
 78   planned_uprate_date                      93588 non-null    datetime64[ns]
 79   previously_canceled                      417936 non-null   boolean       
 80   primary_purpose_id_naics                 6402084 non-null  Int64         
 81   prime_mover_code                         6492408 non-null  string        
 82   pulverized_coal_tech                     342288 non-null   boolean       
 83   reactive_power_output_mvar               3798732 non-null  float64       
 84   retirement_date                          765936 non-null   datetime64[ns]
 85   rto_iso_lmp_node_id                      1350780 non-null  string        
 86   rto_iso_location_wholesale_reporting_id  1113552 non-null  string        
 87   sector_id_eia                            6045024 non-null  Int64         
 88   sector_name_eia                          6169176 non-null  string        
 89   solid_fuel_gasification                  5060580 non-null  boolean       
 90   startup_source_code_1                    520872 non-null   string        
 91   startup_source_code_2                    59130 non-null    string        
 92   startup_source_code_3                    12972 non-null    string        
 93   startup_source_code_4                    4164 non-null     string        
 94   state                                    6534588 non-null  string        
 95   stoker_tech                              62784 non-null    boolean       
 96   street_address                           5925072 non-null  string        
 97   subcritical_tech                         398040 non-null   boolean       
 98   summer_capacity_estimate                 1365708 non-null  boolean       
 99   summer_capacity_mw                       6482033 non-null  float64       
 100  summer_estimated_capability_mw           1033296 non-null  float64       
 101  supercritical_tech                       63432 non-null    boolean       
 102  switch_oil_gas                           1164708 non-null  boolean       
 103  syncronized_transmission_grid            2990220 non-null  boolean       
 104  technology_description                   6296160 non-null  string        
 105  time_cold_shutdown_full_load_code        2117964 non-null  string        
 106  timezone                                 6536712 non-null  string        
 107  topping_bottoming_code                   5481456 non-null  string        
 108  total_fuel_cost                          271010 non-null   float64       
 109  total_mmbtu                              638056 non-null   float64       
 110  turbines_inverters_hydrokinetics         611592 non-null   Int64         
 111  turbines_num                             2266471 non-null  Int64         
 112  ultrasupercritical_tech                  8352 non-null     boolean       
 113  uprate_derate_completed_date             51312 non-null    datetime64[ns]
 114  uprate_derate_during_year                2651676 non-null  boolean       
 115  winter_capacity_estimate                 1305132 non-null  boolean       
 116  winter_capacity_mw                       6483053 non-null  float64       
 117  winter_estimated_capability_mw           1033044 non-null  float64       
 118  zip_code                                 6309984 non-null  string        
dtypes: Int64(9), boolean(28), datetime64[ns](11), float64(26), object(1), string(44)
memory usage: 22.3 GB
CPU times: user 53.9 s, sys: 1.39 s, total: 55.3 s
Wall time: 59.2 s
```

Looking at Xcel Energy's Comanche 3 generator...
```py
comanche3 = mcoe_ms[
    (mcoe_ms.plant_id_eia==470)
    & (mcoe_ms.generator_id=="3")
]
```
It looks like *all* months in the entire time series, even those within years where there's no attribute data available for the generator, are being filled in using the primary key columns, generating a large number of null rows, and inflating the number of rows substantially.  In the cases where we do want "full timeseries" this still seems like it's not what we want -- there's no reason to have entries where there's absolutely no data across decades. The valid range of dates needs to be different for each generator -- if it shows up in the attribute data then it could make sense to include it. But if it doesn't show up in the attribute data at all, we're just making up emptiness. Is this useful / intentional for your use cases @cmgosnell?

I think we've taken the lazy denormalization as far as we can here. 10 million rows and 100+ columns is more than a billion cells in the dataframe, and most of it is information that's being duplicated at least 12 times (for time varying attributes), if not 100 times (for static attributes).  I think at this volume of data being more selective and working with the data as distinct tables with relations between them is the only way things can scale.

katie-lamb commented:
In response to @zaneselvans comment:
> It looks like _all_ months in the entire time series, even those within years where there's no attribute data available for the generator, are being filled in using the primary key columns, generating a large number of null rows, and inflating the number of rows substantially. In the cases where we do want "full timeseries" this still seems like it's not what we want -- there's no reason to have entries where there's absolutely no data across decades. The valid range of dates needs to be different for each generator -- if it shows up in the attribute data then it could make sense to include it. But if it doesn't show up in the attribute data at all, we're just making up emptiness. Is this useful / intentional for your use cases @cmgosnell?
> 
I updated the timeseries expansion so that the dataframe is only expanded to the valid range of dates for each generator, instead of from one start date to one end date for each generator. While this takes out the nulls issue and significantly reduces size, it still blows up memory - I think it was like 5GB for one year of MCOE. So maybe the conclusion is what Zane said previously: 

> I think we've taken the lazy denormalization as far as we can here. 10 million rows and 100+ columns is more than a billion cells in the dataframe, and most of it is information that's being duplicated at least 12 times (for time varying attributes), if not 100 times (for static attributes). I think at this volume of data being more selective and working with the data as distinct tables with relations between them is the only way things can scale.

I think the timeseries expansion in its new form (only expanding between the valid dates for each grouping of ID columns) is still useful, but maybe not appropriate in this context?

zaneselvans commented:
Welp, expanding timeseries only within years for which the generators have data cuts the full MCOE dataframe down to 5.7 million rows, and about 15 GB of memory, so better, but probably still not usable on a machine with 16 GB of memory. But at least the tests can run!

katie-lamb commented:
Current iteration update:
- `expand_timeseries` now fills in until the end of the year
- `mcoe` now only includes specified attributes from the generator table to avoid explosion

There are still a few questions left: 
- Unit tests maybe don't feel right in this case and aren't very clear. Should testing go somewhere else? Should the unit tests just be taken out? Or just formatting them in some way so you don't just need to read the inline DataFrame definitions?
- I added a test for merging quarterly data onto monthly data, but this should probably be updated once the FERC data is actually using this type of quarterly merge. It's kind of acting as a placeholder right now - should I just take it out?

zaneselvans commented:
What is it you don't like about the unit tests? Is it just the mess of having the test dataframes in there inline?

to me this function seems like a good candidate for unit testing generally, since it's pretty generic, and also kind of complex. It would be easy to get something wrong, but it doesn't depend deeply on the details of data being processed in the ETL, so it *seems* like we ought to be able to come up with a modest number of static inputs and expected outputs. to verify that we haven't changed the behavior when it gets edited somehow.

But inline dataframes aren't a lot of fun.  We could also store them as small CSVs in the `tests/data` folder if that would be easier.

katie-lamb commented:
> What is it you don't like about the unit tests? Is it just the mess of having the test dataframes in there inline?
> 
> to me this function seems like a good candidate for unit testing generally, since it's pretty generic, and also kind of complex. It would be easy to get something wrong, but it doesn't depend deeply on the details of data being processed in the ETL, so it _seems_ like we ought to be able to come up with a modest number of static inputs and expected outputs. to verify that we haven't changed the behavior when it gets edited somehow.
> 
> But inline dataframes aren't a lot of fun. We could also store them as small CSVs in the `tests/data` folder if that would be easier.

I stored the most unruly of the inline dataframes as a CSV in the `tests/data` folder. This is the dataframe used for the timeseries fill in test. If this seems good, I can move all of the inline dataframes to CSVs. I figured for the smaller ones, it's nicer to have them inline in one place than spread over CSVs, but they're still not that readable so could go either way.

zaneselvans commented on src/pudl/analysis/mcoe.py:
```diff
@@ -417,6 +402,14 @@ def mcoe(
         all_gens (bool): if True, include attributes of all generators in the
             :ref:`generators_eia860` table, rather than just the generators
             which have records in the derived MCOE values. True by default.
+        all_gens_cols (bool): if True, include all attributes from generators in the
+            :ref:`generators_eia860` table, rather than just the attributes in
+            `default_gens_cols` and the list of extra_gens_cols.
+        extra_gens_cols (List): list of names of column attributes to
+            include from the :ref:`generators_eia860` table in addition to the
```
If you've included type hints in the call signature, they aren't required in the docstring.

zaneselvans commented on src/pudl/helpers.py:
```diff
@@ -332,90 +332,224 @@ def is_doi(doi):
+    return out
+
+
+DateOnCol = Literal["year", "month", "quarter", "day"]
```
Module level constants should `USE_ALL_CAPS` and are typically declared at the beginning of the module... oh but this is a type alias? Did you mean to use this as the type for the `date_on` arguments? They still seem to be Literals. Is that intentional?  You can also annotate constant definitions by including triple-quoted strings on the line following them.

katie-lamb replied:
I wanted the type of the `date_on` column to be a list of strings where the strings must be year, month, quarter, or day. But maybe it's best if the type is just `List[str]` and there's an assert to make sure the strings in `date_on` are valid, as well as being specified in the docstring. I changed this to an assertion, let me know if you think this makes more sense.

zaneselvans commented on src/pudl/output/pudltabl.py:
```diff
@@ -1097,15 +1104,33 @@ def mcoe(
+    def gens_mega_eia(
+        self,
+        update: bool = False,
+        extra_gens_cols=[
```
Using a (mutable) `list` [as a default argument is dangerous](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/), since if the variable is modified in the function, it remains modified for subsequent calls to the function, which can result in unexpected side effects. The recommended solution is to use `None` as the default, check for that value and assign the default value inside the function. You could use an immutable sequence for the default instead, like a `tuple`.

Having two different parameters which define the list of generator attributes to include makes the behavior of the function ambiguous.  What happens if you say `all_gens_cols=True` but also give it set of `extra_gens_cols`?  One solution to this could be to accept either a special literal like `all` or a list of specific columns for `extra_gens_cols` and eliminate the `all_gens_cols` parameter.

Also let's try and add type hints whenever we're touching existing functions or creating new ones.

These comments would also apply to the `mcoe()` function and `plant_parts_eia()` below.

katie-lamb replied:
Good catch, I've never run into this issue of using mutable types as default arguments. Good to know. 

I think a special literal for `extra_gens_cols` to get rid of `all_gens_cols` makes the most sense.

zaneselvans commented on src/pudl/analysis/mcoe.py:
```diff
@@ -474,40 +467,37 @@ def mcoe(
+        "utility_id_pudl",
+        "utility_name_eia",
+    ]
+    if all_gens_cols:
+        gens = pudl_out.gens_eia860()
+    else:
+        gens = pudl_out.gens_eia860()[default_gens_cols + extra_gens_cols]
```
If `extra_gens_cols` can be either `all` or a list of strings (as I suggest below to avoid ambiguity) then the flow control would be:

```py
if extra_gens_cols == "all":
    gens = pudl_out.gens_eia860()
else:
    gens = pudl_out.gens_eia860()[default_gens_cols + extra_gens_cols]
```

zaneselvans commented:
Forgot to mention: I don't feel strongly either way about moving the smaller inline testing dataframes into CSVs or having them remain inline. Whichever you prefer is 👍🏼

katie-lamb commented on test/data/date_merge_unit_test/timeseries_fillin_expected_out.csv:
```diff
@@ -0,0 +1,51 @@
+plant_id_eia,generator_id,report_date,data
+1,1,2/1/19,2
```
OMG ^ this

zaneselvans commented on src/pudl/helpers.py:
```diff
@@ -332,90 +332,224 @@ def is_doi(doi):
+    # equal to Jan 1 of the year after the last real record in the group
+    # this allows records to be filled through the end of the last reported year
+    end_dates = df.groupby(key_cols).agg({date_col: "max"})
+    end_dates.loc[:, date_col] = end_dates[date_col].dt.year + 1
```
Hmm.  There should be some test cases for this probably. We only want to fill in records to the end of the last reported year if we're filling in the time series at a higher frequency. Filling in values for additional years that weren't part of the dataset at all seems like it would cause problems downstream.

katie-lamb replied:
I feel a little confused about desired behavior here. Currently this fills in through the end of the last reported year (see the unit test `test_timeseries_fillin` where monthly data is filled in through December of the last reported year). The only way I could figure out how to do this is to add a dummy row for Jan 1 of the following year. For example:
- last reported value is 2/1/2019
- add dummy row for 1/1/2020
- function fills in monthly values from 2/1/2019 to 12/1/2019
- the dummy row for 1/1/2020 is dropped

It definitely seems like a weird way to do this, but I also couldn't find a nice way to do it so maybe it really is the best way.

 I guess in theory data could be filled in to the end of a specified frequency. So for example if data was daily there could be an argument passed in that specifies that data should be filled in to the end of the month instead of the end of the year.

zaneselvans commented on src/pudl/analysis/mcoe.py:
```diff
@@ -417,6 +402,14 @@ def mcoe(
+        all_gens_cols (bool): if True, include all attributes from generators in the
+            :ref:`generators_eia860` table, rather than just the attributes in
+            `default_gens_cols` and the list of extra_gens_cols.
+        extra_gens_cols (List): list of names of column attributes to
```
A cleaner interface might be:
* Have a generic parameter `gens_cols` that specifies all the columns (or all the data columns) you want returned.
* Use a default value of `None` for `gens_cols` to indicate that a default set of columns should be returned.
* Specify that set of default columns in the top level function and set `gens_cols` to that default value when `None` is passed in, or alternatively have it be a module level constant.
* Use a special keyword `all` to indicate that all columns should be returned, and pass that `all` keyword down until it reaches the `gens_eia860()` call.

katie-lamb commented on src/pudl/analysis/mcoe.py:
```diff
@@ -400,23 +398,29 @@ def mcoe(
 
     Args:
-        pudl_out (pudl.output.pudltable.PudlTabl): a PUDL output object
+        pudl_out (pudl.output.pudltabl.PudlTabl): a PUDL output object
```
I moved all of the type hints to the call signature in response to @zaneselvans  [comment ](https://github.com/catalyst-cooperative/pudl/pull/1550#discussion_r883871312). However moving this to the call signature caused a circular import issue, so this one I left in the docstring. Not sure if it's better to have consistency and just leave out the type hint for `pudl_out`.

cmgosnell replied:
I've also had the same circular reference issue :-( i don't understand what the correct implementation is. regardless, I think this is fine solution even if its not consistent

katie-lamb commented on src/pudl/analysis/mcoe.py:
```diff
@@ -1,9 +1,23 @@
+    "utility_id_eia",
+    "utility_id_pudl",
+    "utility_name_eia",
+]
```
This used to be a variable in the `mcoe` function. I decided to make it a global because it feels like there's really no good place for it and I want it to be more accessible / viewable from outside the `mcoe` function. I tried to just do away with default generator columns as a whole but these ones still feel pretty essential, especially since `plant_id_eia`, `generator_id` , and `report_date` are all necessary for the merge.

cmgosnell replied:
I think this a good overall solution for this!

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -332,90 +332,271 @@ def is_doi(doi):
+    key_cols: List[str],
+    date_col: str = "report_date",
+    freq: str = "MS",
+    fill_through_freq: Literal["year", "month", "day"] = "year",
```
This argument was added so now you can fill through the end of a desired time frequency, not just year. I think this was what was being discussed [in this thread](https://github.com/catalyst-cooperative/pudl/pull/1550#discussion_r883931974). This makes the whole thing more confusing and less clean so if this functionality doesn't seem necessary maybe it should get taken out. I also added a test for filling through the end of the month.

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -332,90 +332,271 @@ def is_doi(doi):
+    # desired fill_through_freq.
+    # This allows records to be filled through the end of the last reported period
+    # and then this dummy record is dropped
+    df = convert_col_to_datetime(df, date_col)
```
I thinkkk this `convert_col_to_datetime` could just happen on `end_dates` (after the `groupby` call one line later) and it would be faster. But this seemed safer? In case report date strings were formatted weirdly and `groupby` + `max` didn't get the latest date?

katie-lamb commented on src/pudl/helpers.py:
```diff
@@ -332,90 +332,271 @@ def is_doi(doi):
+        if col not in ["year", "month", "quarter", "day"]:
+            raise AssertionError(
+                logger.error(f"{col} is not a valid string in date_on column list.")
+            )
```
This assert was put in in response to the issue with type hints and `Literal["year", "month", "quarter", "day"]` (see [this thread](https://github.com/catalyst-cooperative/pudl/pull/1550#discussion_r883874719)).

katie-lamb commented:
Latest update: I think the code is close! I'm pretty sure I addressed all comments from @zaneselvans and @cmgosnell . 

I'm going to take a look at how the `date_merge` changed the mcoe table from when the `clean_merge_asof` was being performed. I also am going to look at how much / if the `date_merge` sans timeseries fill in spend up the merge.

katie-lamb commented on src/pudl/analysis/mcoe.py:
```diff
@@ -474,40 +478,28 @@ def mcoe(
+    else:
+        gens = pudl_out.gens_eia860()[list(set(DEFAULT_GENS_COLS + gens_cols))]
+
+    mcoe_out = pudl.helpers.full_timeseries_date_merge(
```
Should there be a parameter that allows you to generate the MCOE table without the time series fill in and have this just be `date_merge`?

katie-lamb replied:
I went ahead and added this parameter. Lmk if it doesn't make sense. It defaults to just do the `date_merge` not the `full_timeseries_date_merge`.

katie-lamb commented on test/validate/mcoe_test.py:
```diff
@@ -108,11 +108,11 @@ def test_no_null_rows_mcoe(pudl_out_mcoe, live_dbs, df_name, thresh):
-        ("fuel_cost", 514_642, 42_958),
-        ("capacity_factor", 559_570, 46_718),
-        ("mcoe", 559_618, 46_718),
+        ("hr_by_unit", 341_640, 28_519),
+        ("hr_by_gen", 514_619, 42_935),
+        ("fuel_cost", 514_619, 42_935),
+        ("capacity_factor", 559_546, 46_694),
+        ("mcoe", 559_594, 46_694),
```
See [this explanation](https://github.com/catalyst-cooperative/pudl/pull/1550#issuecomment-1146456525) for why `date_merge` produces fewer rows.

katie-lamb commented:
@zaneselvans  brings up that the missing utility info for certain years (referenced in [this comment](https://github.com/catalyst-cooperative/pudl/pull/1550#issuecomment-1146456525)) is weird.

These are a few of the plant IDs that are missing utility data in `plant_utils_eia860` for a year. They are alll 2010 in this screenshot but there are also some plants missing 2012 and 2009 utility data. Honestly I couldn't figure out what the best way to reference the full CSV was in this comment but I have the full list of plants and the years they're missing utility data.

<img width="847" alt="Screen Shot 2022-06-08 at 11 47 10 AM" src="https://user-images.githubusercontent.com/95320980/172693368-56bea0f4-b73a-4eb1-89e9-e4d6da031d92.png">

cmgosnell commented on src/pudl/output/pudltabl.py:
```diff
@@ -1127,32 +1146,64 @@ def gens_mega_eia(self, update: bool = False) -> pd.DataFrame:
                     "The frequency of the pudl_out object must be `AS` for the "
                     f"plant-parts table and we got {self.freq}"
                 )
+            if gens_cols is None:
```
nice! I like seeing this `gens_cols` in action, should we require these columns to be passed in no matter what?

cmgosnell commented on src/pudl/output/pudltabl.py:
```diff
@@ -1127,32 +1146,64 @@ def gens_mega_eia(self, update: bool = False) -> pd.DataFrame:
+                    "operational_status",
+                    "capacity_mw",
+                    "fuel_type_code_pudl",
+                    "planned_retirement_date",
```
why are these last two columns different than the columns in `gens_mega_eia`? because these two are needed for the full ppl but aren't for the gens_mega?

katie-lamb replied:
yesss exactly

cmgosnell reviewed (approved):
it looks good to me! I'll be honest - the more i look at it the more I am lightly confused about what changes happened when. but overall it looks great and i am excited to get this integrated 🎉

zaneselvans reviewed (approved):
I just can't take it any more 🙈 🙉 😬
